File: PathoFroh/src/main/java/com/patho/main/model/interfaces/FullName.java
package com.patho.main.model.interfaces;

import javax.persistence.Transient;

import com.patho.main.model.Person.Gender;

/**
 * Interface for returning a merged name from first, last name and title.
 * 
 * @author andi
 *
 */
public interface FullName  {

	public String getTitle();

	public String getFirstName();

	public String getLastName();

	public default Gender getGender() {
		return Gender.UNKNOWN;
	}

	/**
	 * Returns a full name with title, name and surname.
	 * 
	 * @return
	 */
	@Transient
	public default String getFullName() {
		StringBuilder result = new StringBuilder();

		if (getTitle() != null && !getTitle().isEmpty())
			result.append(getTitle() + " ");

		if (getFirstName() != null && !getFirstName().isEmpty())
			result.append(getFirstName() + " ");

		if (getLastName() != null && !getLastName().isEmpty())
			result.append(getLastName() + " ");

		// remove the last space from the string
		if (result.length() > 0)
			return result.substring(0, result.length() - 1);
		else
			return "";
	}

	/**
	 * Returns a title + name, if no title is provided,
	 * 
	 * @return
	 */
	@Transient
	public default String getFullNameAndTitle() {
		StringBuilder result = new StringBuilder();

		if (getTitle() != null && !getTitle().isEmpty())
			result.append(getTitle() + " ");
		else {
			// TODO hardcoded!
			if (getGender() == Gender.FEMALE)
				result.append("Frau ");
			else
				result.append("Herr ");
		}

		int index = result.indexOf("Apl.");
		if (index != -1)
			result.replace(index, 4, "");

		if (getLastName() != null && !getLastName().isEmpty())
			result.append(getLastName() + " ");

		return result.substring(0, result.length() - 1);
	}
}

File: src/main/kotlin/com/patho/main/model/interfaces/FullName.kt
package com.patho.main.model.interfaces

import com.patho.main.model.person.Person
import com.patho.main.service.impl.SpringContextBridge
import javax.persistence.Transient

interface FullName {

    var title: String

    var firstName: String

    var lastName: String

    var gender: Person.Gender

    /**
     * Returns a full name with title, name and surname.
     *
     * @return
     */
    @Transient
    fun getFullName(): String {
        val result = StringBuilder()

        if (title.isNotEmpty()) result.append("$title ")
        if (firstName.isNotEmpty()) result.append("$firstName ")
        if (lastName.isNotEmpty()) result.append("$lastName ")

        // remove the last space from the string
        return if (result.isNotEmpty())
            result.substring(0, result.length - 1)
        else
            ""
    }

    /**
     * Returns a gender name, surname and title
     *
     * @return
     */
    @Transient
    fun getFullNameAndTitle(): String {
        val result = StringBuilder()
        if (gender == Person.Gender.FEMALE)
            result.append(SpringContextBridge.services().resourceBundle.get("enum.gender.FEMALE_TITLE"))
        else
            result.append(SpringContextBridge.services().resourceBundle.get("enum.gender.MALE_TITLE"))

        result.append(getFullName())

        return result.substring(0, result.length - 1).replace("Apl.", "")
    }

    /**
     * Returns gender name and title
     */
    fun getNameAndTitle(): String {
        val result = StringBuilder()

        if (gender == Person.Gender.FEMALE)
            result.append(SpringContextBridge.services().resourceBundle.get("enum.gender.FEMALE_TITLE"))
        else
            result.append(SpringContextBridge.services().resourceBundle.get("enum.gender.MALE_TITLE"))

        if (title.isNotEmpty()) result.append("$title ")
        if (lastName.isNotEmpty()) result.append("$lastName ")

        return result.substring(0, result.length - 1).replace("Apl.", "")
    }
}